packages/collector/src/announceCycle/unannounced.js

```diff
@@ -64,43 +91,106 @@ function tryToAnnounce(ctx, retryDelay = initialRetryDelay) {
+        agentResponse.secrets.list
+      );
+    } else {
+      secrets.setMatcher(agentResponse.secrets.matcher, agentResponse.secrets.list);
     }
-    if (response.spanBatchingEnabled === true || response.spanBatchingEnabled === 'true') {
-      logger.info('Enabling span batching via agent configuration.');
-      tracing.enableSpanBatching();
-    }
+  }
+}
 
-    if (response.secrets) {
-      if (!(typeof response.secrets.matcher === 'string')) {
-        logger.warn(
-          'Received invalid secrets configuration from agent, attribute matcher is not a string: $s',
-          response.secrets.matcher
-        );
-      } else if (Object.keys(secrets.matchers).indexOf(response.secrets.matcher) < 0) {
-        logger.warn(
-          'Received invalid secrets configuration from agent, matcher is not supported: $s',
-          response.secrets.matcher
-        );
-      } else if (!Array.isArray(response.secrets.list)) {
-        logger.warn(
-          'Received invalid secrets configuration from agent, attribute list is not an array: $s',
-          response.secrets.list
-        );
-      } else {
-        secrets.setMatcher(response.secrets.matcher, response.secrets.list);
-      }
-    }
+/**
+ * @param {AgentAnnounceResponse} agentResponse
+ */
+function applyExtraHttpHeaderConfiguration(agentResponse) {
+  if (agentResponse.tracing) {
+    actuallyApplyExtraHttpHeaderConfiguration(agentResponse.tracing['extra-http-headers']);
+    return;
+  }
 
-    ctx.transitionTo('announced');
-  });
+  // Fallback for Node.js discovery prior to version 1.2.18, which did not support providing the complete
+  // com.instana.tracing section in the agent response. We can remove this fallback approximately in May 2023.
+  actuallyApplyExtraHttpHeaderConfiguration(agentResponse.extraHeaders);
+}
+
+/**
+ * @param {Array.<string>} extraHeaders
+ */
+function actuallyApplyExtraHttpHeaderConfiguration(extraHeaders) {
+  if (Array.isArray(extraHeaders)) {
+    tracing.setExtraHttpHeadersToCapture(
+      // The Node.js HTTP API converts all incoming HTTP headers into lowercase.
+      extraHeaders.map((/** @type {string} */ s) => s.toLowerCase())
+    );
+  }
+}
+
+/**
+ * @param {AgentAnnounceResponse} agentResponse
+ */
+function applyKafkaTracingConfiguration(agentResponse) {
+  if (agentResponse.tracing && agentResponse.tracing.kafka) {
+    const kafkaTracingConfigFromAgent = agentResponse.tracing.kafka;
+    const kafkaTracingConfig = {
+      traceCorrelation:
+        kafkaTracingConfigFromAgent['trace-correlation'] != null
+          ? kafkaTracingConfigFromAgent['trace-correlation']
+          : tracingConstants.kafkaTraceCorrelationDefault,
+      headerFormat:
+        kafkaTracingConfigFromAgent['header-format'] != null
+          ? kafkaTracingConfigFromAgent['header-format']
+          : tracingConstants.kafkaHeaderFormatDefault
+    };
+    tracing.setKafkaTracingConfig(kafkaTracingConfig);
```

Comment:
_Suggestion_:

The code to override the config from the agent for the target instrumentation is a bit hard to follow, but of course it existed before this PR :)

As far as I can see, we do not trace before the announceCycle is completed and reaches `agentReady` step, which calls `tracing.activate`.

In the `unannounced` step we try to call the core package (which I think is unneeded and too coupled) to tell the instrumentation that there is another config available, which is more important.

IMO the unannounced cycle step should write the agent config into the ctx object. And when the last step is reached we pass `tracing.activate(ctx.agentConfig)`. We then automatically pass the agent config into each instrumentation. Each instrumentation can pick the config it needs.

Advantage: 
- we do not have to iterate 3 times through all instrumentations (costs time)
- easier to read
- we don't need setExtraHttpHeadersToCapture, setKafkaTracingConfig or similar. If an instrumentation needs config from agent, it can just access the agentConfig in the activate fn

Happy to discuss and refactor afterwards.

Reply:
I like this idea! I'll look into that.

Reply:
I have implemented this suggestion now with the small difference that I attached the configuration that the agent has provided to agent/opts, which already was used to store the agent uuid, so it felt somewhat natural to put the config there, too.

I know you already approved but somehow this triggered a couple more changes (deleting an obsolete test) so I would actually appreciate a re-review of the latest changes if you don't mind :-)